pkg/declarative/v2/options.go

```diff
@@ -0,0 +1,285 @@
+}
+
+type Option interface {
+	Apply(options *Options)
+}
+
+func (o *Options) Apply(options ...Option) *Options {
+	for i := range options {
+		options[i].Apply(o)
+	}
+	return o
+}
+
+type WithNamespaceOption struct {
+	name            string
+	createIfMissing bool
+}
+
+func WithNamespace(name string, createIfMissing bool) WithNamespaceOption {
+	return WithNamespaceOption{
+		name:            name,
+		createIfMissing: createIfMissing,
+	}
+}
+
+func (o WithNamespaceOption) Apply(options *Options) {
+	options.Namespace = o.name
+	options.CreateNamespace = o.createIfMissing
+}
+
+type WithFieldOwner client.FieldOwner
+
+func (o WithFieldOwner) Apply(options *Options) {
+	options.FieldOwner = client.FieldOwner(o)
+}
+
+type WithFinalizer string
+
+func (o WithFinalizer) Apply(options *Options) {
+	options.Finalizer = string(o)
+}
+
+type WithManagerOption struct {
+	manager.Manager
+}
+
+func WithManager(mgr manager.Manager) WithManagerOption {
+	return WithManagerOption{Manager: mgr}
+}
+
+func (o WithManagerOption) Apply(options *Options) {
+	options.EventRecorder = o.GetEventRecorderFor(EventRecorderDefault)
+	options.Config = o.GetConfig()
+	options.Client = o.GetClient()
+}
+
+type WithCustomResourceLabels labels.Set
+
+func (o WithCustomResourceLabels) Apply(options *Options) {
+	labelTransform := func(ctx context.Context, object Object, resources []*unstructured.Unstructured) error {
+		for _, targetResource := range resources {
+			lbls := targetResource.GetLabels()
+			if lbls == nil {
+				lbls = labels.Set{}
+			}
+			for s := range o {
+				lbls[s] = o[s]
+			}
+			targetResource.SetLabels(lbls)
+		}
+		return nil
+	}
+	options.PostRenderTransforms = append(options.PostRenderTransforms, labelTransform)
+}
+
+func WithSpecResolver(resolver SpecResolver) SpecResolverOption {
+	return SpecResolverOption{resolver}
+}
+
+type SpecResolverOption struct {
+	SpecResolver
+}
+
+func (o SpecResolverOption) Apply(options *Options) {
+	options.SpecResolver = o
+}
+
+type ObjectTransform = func(context.Context, Object, []*unstructured.Unstructured) error
+
+func WithPostRenderTransform(transforms ...ObjectTransform) PostRenderTransformOption {
+	return PostRenderTransformOption{transforms}
+}
+
+type PostRenderTransformOption struct {
+	ObjectTransforms []ObjectTransform
+}
+
+func (o PostRenderTransformOption) Apply(options *Options) {
+	options.PostRenderTransforms = append(options.PostRenderTransforms, o.ObjectTransforms...)
+}
+
+// Hook defines a Hook into the declarative reconciliation
+// skr is the runtime cluster
+// kcp is the control-plane cluster
+// obj is guaranteed to be the reconciled object and also to always preside in kcp.
+type Hook func(ctx context.Context, skr Client, kcp client.Client, obj Object) error
+
+type (
+	PostRun   Hook
+	PreDelete Hook
+)
+
+type WithPostRun []PostRun
+
+func (o WithPostRun) Apply(options *Options) {
+	options.PostRuns = append(options.PostRuns, o...)
+}
+
+type WithPreDelete []PreDelete
```

Comment:
Should we add a comment to tell `WithPostRun` and `WithPreDelete` is only for internal use?

Reply:
+1. I think we can add a warning that this can lead to unintended side effects if used incorrectly.